pyomo/contrib/doe/fim_doe.py

```diff
@@ -6,13 +6,127 @@
+        '''
+        measurement_name = list(all_info.keys())
+        # a list of measurement extra indexes
+        measurement_extra_index = []
+        # a list of measurement names with extra indexes
+        extra_measure_name = []
+        # check if the measurement has extra indexes
+        for i in measurement_name:
+            if type(all_info[i]) is dict:
+                index_list = list(all_info[i].keys())
+                extra_measure_name.append(i)
+                measurement_extra_index.append(index_list)
+            elif type(all_info[i]) is list:
+                measurement_extra_index.append(None)
+        # a dictionary, keys are measurement names, values are a list of extra indexes
+        name_and_index = {}
+        for i, iname in enumerate(measurement_name):
+            name_and_index[iname] = measurement_extra_index[i]
+
+        self.name_and_index = name_and_index
+
+    def __generate_flatten_name(self, measure_name_and_index):
+        '''Generate measurement flattened names
+        Parameters
+        ----------
+        measure_name_and_index: a dictionary, keys are measurement names, values are lists of extra indexes
+
+        Returns
+        ------
+        jac_involved_name: a list of flattened measurement names
+        '''
+        flatten_names = []
+        for j in list(measure_name_and_index.keys()):
+            if measure_name_and_index[j] is not None: # if it has extra index
+                for ind in measure_name_and_index[j]:
+                    flatten_name = j + '_index_' + str(ind)
+                    flatten_names.append(flatten_name)
+            else:
+                flatten_names.append(j)
+
+        self.flatten_measure_name = flatten_names
+
+    def __generate_flatten_timeset(self, all_info, flatten_measure_name,name_and_index):
+        '''
+        Generate flatten variables timeset. Return a dict where keys are the flattened variable names,
+        values are a list of measurement time.
+        '''
+        flatten_measure_timeset = {}
+        for i in flatten_measure_name:
+            # split the flattened name if needed
+            if '_index_' in i:
+                measure_name = i.split('_index_')[0]
+                measure_index = i.split('_index_')[1]
+                if type(name_and_index[measure_name][0]) is int:
+                    measure_index = int(measure_index)
+                flatten_measure_timeset[i] = all_info[measure_name][measure_index]
+            else:
+                flatten_measure_timeset[i] = all_info[i]
+        self.flatten_measure_timeset = flatten_measure_timeset
+
+    def check_subset(self,subset, throw_error=True):
+        '''
+        Check if the subset is correctly defined with right name, index and time.
+        subset: measurement name and index involved in jacobian calculation
+        throw_error: if the given subset is not a subset of the measurement set, throw error message
+        '''
```

Comment:
Add to docstring.

Arguments:
    subset: description here

Returns:
    valid_subset: Boolean, true if `subset` is contained in `self`